h/static/scripts/directive/test/annotation-test.js

```diff
@@ -0,0 +1,1377 @@
+      annotation.id = null;
+      createDirective();
+      fakeGroups.focused = sinon.stub().returns({
+        id: 'new-group'
+      });
+      $rootScope.$broadcast(events.GROUP_FOCUSED);
+      assert.equal(annotation.group, 'new-group');
+    });
+  });
+
+  it(
+    'updates perms when moving new annotations to the focused group',
+    function() {
+      // id must be null so that AnnotationController considers this a new
+      // annotation.
+      annotation.id = null;
+      annotation.group = 'old-group';
+      annotation.permissions = {
+        read: [annotation.group]
+      };
+      // This is a shared annotation.
+      fakePermissions.isShared.returns(true);
+      createDirective();
+      // Make permissions.shared() behave like we expect it to.
+      fakePermissions.shared = function(groupId) {
+        return {
+          read: [groupId]
+        };
+      };
+      fakeGroups.focused = sinon.stub().returns({
+        id: 'new-group'
+      });
+      $rootScope.$broadcast(events.GROUP_FOCUSED);
+      assert.deepEqual(annotation.permissions.read, ['new-group']);
+    }
+  );
+
+  it('saves shared permissions for the new group to drafts', function() {
+    // id must be null so that AnnotationController considers this a new
+    // annotation.
+    annotation.id = null;
+    annotation.group = 'old-group';
+    annotation.permissions = {
+      read: [annotation.group]
+    };
+    // This is a shared annotation.
+    fakePermissions.isShared.returns(true);
+    createDirective();
+    // drafts.get() needs to return something truthy, otherwise
+    // AnnotationController won't try to update the draft for the annotation.
+    fakeDrafts.get.returns(true);
+    // Make permissions.shared() behave like we expect it to.
+    fakePermissions.shared = function(groupId) {
+      return {
+        read: [groupId]
+      };
+    };
+    // Change the focused group.
+    fakeGroups.focused = sinon.stub().returns({
+      id: 'new-group'
+    });
+    $rootScope.$broadcast(events.GROUP_FOCUSED);
+    assert.deepEqual(
+      fakeDrafts.update.lastCall.args[1].permissions.read,
+      ['new-group'],
+      'Shared permissions for the new group should be saved to drafts');
+  });
+
+  it('does not change perms when moving new private annotations', function() {
+    // id must be null so that AnnotationController considers this a new
+    // annotation.
+    annotation.id = null;
+    annotation.group = 'old-group';
+    annotation.permissions = {
+      read: ['acct:bill@localhost']
+    };
+    createDirective();
+    // This is a private annotation.
+    fakePermissions.isShared.returns(false);
+    fakeGroups.focused = sinon.stub().returns({
+      id: 'new-group'
+    });
+    $rootScope.$broadcast(events.GROUP_FOCUSED);
+    assert.deepEqual(
+      annotation.permissions.read, ['acct:bill@localhost'],
+      'The annotation should still be private');
+  });
+});
+
+describe('AnnotationController', function() {
```

Comment:
The solution to this I think is simply to have smaller units with fewer dependencies.